File: p9012.cpp
#include<bits/stdc++.h>
using namespace std;
int main(){
    int T;
    cin >> T;
    vector<bool> answer;
    for(int i = 0; i < T; i++){
        string input;
        cin >> input;

        stack<char> st;
        for(int j = 0; j < input.size(); j++){
            if(input[j] == ')'){
                if(st.empty()){
                    st.push(')');
                    break;
                }
                else{
                    st.pop();
                    continue;
                }
            }
            st.push(input[j]);
        }

        answer.push_back(st.empty());
    }

    string a;
    for(bool ans: answer){
        a = ans ? "YES" : "NO";
        cout << a << "\n";
    }
}

File: p9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class p9012 {
    public static int tc;
    public static String input;
    public static Stack<Character> st;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        tc = Integer.parseInt(br.readLine());
        while(tc-- > 0){
            st = new Stack<>();
            input = br.readLine();
            String res = "YES";
            for(int i = 0; i < input.length(); i++){
                if(input.charAt(i) == ')') {
                    if(!st.empty() && st.peek() == '('){
                        st.pop();
                    }
                    else{
                        res = "NO";
                        break;
                    }
                }
                else st.push(input.charAt(i));
            }
            if(!st.empty()) res = "NO";
            System.out.println(res);
        }
    }
}
